src/main/java/io/deepstream/utils/AckTimeoutRegistry.java

```diff
@@ -0,0 +1,167 @@
+package io.deepstream.utils;
+
+import io.deepstream.ConnectionChangeListener;
+import io.deepstream.IDeepstreamClient;
+import io.deepstream.constants.Actions;
+import io.deepstream.constants.ConnectionState;
+import io.deepstream.constants.Event;
+import io.deepstream.constants.Topic;
+import io.deepstream.message.Message;
+
+import java.util.*;
+import java.util.concurrent.LinkedBlockingQueue;
+import java.util.TimerTask;
+
+public class AckTimeoutRegistry implements ConnectionChangeListener {
+
+    private Map<String, AckTimeoutTask> register;
+    private Timer timer;
+    private Topic topic;
+    private long timeoutDuration;
+    private IDeepstreamClient client;
+    private ConnectionState state;
+    private LinkedBlockingQueue<AckTimeoutTask> ackTimers;
+
+    /**
+     * The registry for all ack timeouts.
+     *
+     * @param client The client it sends errors to
+     * @param topic The topic for all acks in the register
+     * @param timeoutDuration The timeout duration for acks in milliseconds
+     */
+    public AckTimeoutRegistry( IDeepstreamClient client, Topic topic, long timeoutDuration ) {
+        this.client = client;
+        this.register = new HashMap<String, AckTimeoutTask>();
+        this.ackTimers = new LinkedBlockingQueue<AckTimeoutTask>();
+        this.timer = new Timer();
+        this.topic = topic;
+        this.timeoutDuration = timeoutDuration;
+
+        this.state = client.getConnectionState();
+        this.client.addConnectionChangeListener( this );
+    }
+
+    /**
+     * Clears the ack timeout for a message.
+     *
+     * @param message The message received to remove the ack timer for
+     */
+    public void clear( Message message ) {
+        String uniqueName = this.getUniqueName( Actions.getAction( message.data[ 0 ] ), message.data[ 1 ] );
+        if( this.clear( uniqueName ) == false ) {
+            this.client.onError( this.topic, Event.UNSOLICITED_MESSAGE, message.raw );
+        }
+    }
+
+    /**
+     * Checks to see if an ack timer already exists in the register for
+     * the given name. If it does, it clears it, then starts a new one.
+     *
+     * @param name The name to be added to the register
+     *
+     * @deprecated This is flagged as deprecated since we should avoid using it ( but it's a JS port
+     * so is worth flagging when used ).
+     */

```

Comment:
As mentioned in docs, I'm not certain if we want to have this API, since every message send will have a TOPIC and it's possibly there for backwards compatibility in the JS client. You in favour of deleting it? I didn't since we had tests and I felt guilty tbh :$